sw/device/silicon_creator/lib/sigverify/sphincsplus/params.h

```diff
@@ -0,0 +1,149 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Derived from code in the SPHINCS+ reference implementation (CC0 license):
+// https://github.com/sphincs/sphincsplus/blob/ed15dd78658f63288c7492c00260d86154b84637/ref/params/params-sphincs-shake-128s.h
+
+#ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_SIGVERIFY_SPHINCSPLUS_PARAMS_H_
+#define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_SIGVERIFY_SPHINCSPLUS_PARAMS_H_
+
+/**
+ * This file represents the SPHINCS+ parameter set shake-128s, meaning:
+ * - The hash function is SHAKE-256
+ * - >= 128 bits of security for up to 2^64 signatures
+ * - The parameter set is optimized to be small "s" rather than fast "f"
+ *   - The "fast" variant is faster for signing but actually slower for
+ *     verification, so "s" is a better fit given we don't care about signing
+ *     speed.
+ *
+ * For more details on parameterization, see the SPHINCS+ paper:
+ * https://sphincs.org/data/sphincs+-paper.pdf
+ */
+
+enum {
+  /**
+   * Hash output length in bytes.
+   */
+  kSpxN = 16,
+  /**
+   * Height of the hypertree.
+   */
+  kSpxFullHeight = 63,
+  /**
+   * Number of subtree layers.
+   */
+  kSpxD = 7,
+  /**
+   * FORS tree dimension (height).
+   */
+  kSpxForsHeight = 12,
+  /**
+   * FORS tree dimension (number of trees).
+   */
+  kSpxForsTrees = 14,
+  /**
+   * Winternitz parameter.
+   */
+  kSpxWotsW = 16,
+  /**
+   * Number of bytes in a hypertree address (for clarity).
+   */
+  kSpxAddrBytes = 32,
+  /**
+   * Bit-length of the Winternitz parameter.
+   */
+  kSpxWotsLogW = 4,
+  /**
+   * Parameter `len1` for WOTS signatures.
+   *
+   * See section 3.1 of the SPHINCS+ NIST submission:
+   *   https://sphincs.org/data/sphincs+-r3.1-specification.pdf
+   */
+  kSpxWotsLen1 = (8 * kSpxN) / kSpxWotsLogW,
+  /**
+   * Parameter `len2` for WOTS signatures.
+   *
+   * This value is precomputed and equal to:
+   *   floor(log(len1 * (w - 1)) / log(w)) + 1
+   *
+   * During the WOTS computation, the maximum checksum value is `len1 * (w -
+   * 1)`. The log() functions here essentially apply the change-of-base rule;
+   * what we are actually computing is floor(log_w(len1 * (w - 1))) + 1, which
+   * expresses the number of base-w integers required to encode the checksum.
+   *
+   * Some precomputed values based on w and n:
+   *   +------+----------------+----------+
+   *   |  w   |        n       |   len2   |
+   *   +------+----------------+----------+
+   *   | 256  |   0 < n <= 1   |    1     |
+   *   | 256  |   1 < n <= 256 |    2     |
+   *   |  16  |   0 < n <= 8   |    2     |
+   *   |  16  |   8 < n <= 136 |    3     |
+   *   |  16  | 136 < n <= 256 |    4     |
+   *   +------+----------------+----------+
+   *
+   * See section 3.1 of the SPHINCS+ NIST submission:
+   *   https://sphincs.org/data/sphincs+-r3.1-specification.pdf
+   */
+  kSpxWotsLen2 = 4,
```

Comment:
nit: When we start writing unit tests for spx+, we should add a simple unit test for verifying that this is indeed `floor(log(len1 * (w - 1)) / log(w)) + 1`.

Reply:
Agreed -- I had some `static_assert`s for this but for some reason the compiler didn't like them. I'll add a unit test file for the parameters that includes all the things I wanted to put in `static_assert`.